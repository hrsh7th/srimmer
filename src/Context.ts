import * as equals from "shallowequal";

/**
 * Mutate function.
 */
export type Update<State> = (state: State) => void;

/**
 * Select function.
 */
export type Select<State, T = any> = (state: State) => T;

/**
 * bit's state.
 */
export type BitState = {
  bit: number;
  version: number;
  state?: any;
};

/**
 * Context.
 */
export class Context<State> {
  /**
   * get state function(Provider#getState).
   */
  public getState?: () => State;

  /**
   * update state function(Provider#updateState).
   */
  public updateState?: (update: Update<State>) => void;

  /**
   * Consumer's select props as bitmask.
   */
  private bitmask: Map<Select<State>, BitState> = new Map();

  /**
   * calculate bitmask.
   */
  public calculateBitmask(nextState: State) {
    return Array.from(this.bitmask.entries()).reduce(
      (bits, [select, bitState]) => {
        const next = select(nextState);
        const isSame = equals.default(next, bitState.state);
        if (!isSame) {
          bitState.state = next;
          bitState.version++;
          return bits | bitState.bit;
        }
        return bits;
      },
      0
    );
  }

  /**
   * register select function.
   */
  public register(select: Select<State>) {
    this.bitmask.set(select, {
      bit: this.bitmask.size + 1,
      version: 1
    });
  }

  /**
   * getBitState.
   */
  public getBitState(select: (state: State) => any) {
    return this.bitmask.get(select)!;
  }
}
