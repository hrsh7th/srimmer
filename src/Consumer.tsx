import React from "react";
import { Context, Select } from "./Context";

export type ConsumerProps<State, T = any> = {
  select: Select<State, T>;
  children: (selected: T) => React.ReactNode;
};

export function createConsumer<State>(
  Consumer: React.Consumer<State>,
  context: Context<State>
) {
  /**
   * SrimerConsumer.
   */
  return class SrimmerConsumer<T> extends React.Component<
    ConsumerProps<State, T>
  > {
    /**
     * initialize bitmask.
     */
    public constructor(props: ConsumerProps<State>) {
      super(props);
      context.register(props.select);
    }

    /**
     * rednder.
     */
    public render() {
      return <Consumer>{this.select}</Consumer>;
    }

    /**
     * select and versioning.
     */
    public select = (state: State) => {
      const bitState = context.getBitState(this.props.select);
      bitState.state = bitState.state || this.props.select(state);
      return (
        <Versioning version={bitState.version}>
          {() => this.props.children(bitState.state)}
        </Versioning>
      );
    };
  };
}

/**
 * Versioning selected state to avoid re-rendering.
 */
class Versioning extends React.Component<{
  version: number;
  children: () => React.ReactNode;
}> {
  public shouldComponentUpdate(props: { version: number }) {
    return this.props.version !== props.version;
  }

  public render() {
    return this.props.children();
  }
}
