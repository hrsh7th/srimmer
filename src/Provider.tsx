import React from "react";
import { produce } from "immer";
import { Context, Update } from "./Context";

export type ProviderProps<State> = {
  state: State;
  children: React.ReactNode;
};

export type ProviderState<State> = {
  state: State;
};

export function createProvider<State>(
  Provider: React.Provider<State>,
  context: Context<State>
) {
  /**
   * SrimmerProvider.
   */
  return class SrimmerProvider extends React.Component<
    ProviderProps<State>,
    ProviderState<State>
  > {
    /**
     * props to state.
     */
    public constructor(props: ProviderProps<State>) {
      super(props);
      this.state = { state: props.state };
      context.getState = this.getState;
      context.updateState = this.updateState;
    }

    /**
     * render.
     */
    public render() {
      return (
        <Provider value={this.state.state}>{this.props.children}</Provider>
      );
    }

    /**
     * state getter.
     */
    public getState = () => {
      return this.state.state;
    };

    /**
     * state updator.
     */
    public updateState = (update: Update<State>) => {
      this.setState({
        state: produce<State>(this.state.state, function(state) {
          update(state as State);
        })
      });
    };
  };
}
