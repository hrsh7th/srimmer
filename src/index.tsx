import React from "react";
import { PatchListener } from "immer";
import { Context, Select } from "./Context";
import { createConsumer } from "./Consumer";
import { createProvider } from "./Provider";

export type Select<
  T extends React.ComponentType<{ children: any }>
> = T extends React.ComponentType<{
  children: (selected: infer R) => React.ReactNode;
}>
  ? R
  : never;

export function define<State extends object>() {
  const context = new Context<State>();
  const { Provider, Consumer } = React.createContext<State>({} as any);
  const SrimmerProvider = createProvider<State>(Provider, context);
  const SrimmerConsumer = createConsumer<State>(Consumer, context);

  return {
    Provider: SrimmerProvider,
    Consumer: SrimmerConsumer,
    get: context.getState,
    set: context.setState,
    update: context.updateState,
    patches: (patchListener: PatchListener) => {
      context.patchListener = patchListener;
    },
    select: function select<T>(select: Select<State, T>) {
      return (props: { children: (s: T) => React.ReactNode }) => (
        <SrimmerConsumer<T> select={select}>{props.children}</SrimmerConsumer>
      );
    }
  };
}
