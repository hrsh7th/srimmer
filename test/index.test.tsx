import * as React from "react";
import { create } from "react-test-renderer";
import { define } from "../src";

let state: {
  consumer1: { value: number };
  consumer2: { value: number };
};
type State = typeof state;

beforeAll(() => {
  state = {
    consumer1: { value: 0 },
    consumer2: { value: 0 }
  };
});

test("Should re-rendering only part of consumers.", () => {
  const { Provider, update, select } = define<State>();
  const renderCount = { consumer1: 0, consumer2: 0 };
  const Consumer1 = select(state => state.consumer1);
  const Consumer2 = select(state => state.consumer2);
  const component = create(
    <Provider state={state}>
      <Consumer1>
        {state => (
          <div>
            {String(state.value)}
            {String(renderCount.consumer1++)}
          </div>
        )}
      </Consumer1>
      <Consumer2>
        {state => (
          <div>
            {String(state.value)}
            {String(renderCount.consumer2++)}
          </div>
        )}
      </Consumer2>
    </Provider>
  );
  update(state => state.consumer1.value++);
  expect(component.toJSON()).toMatchSnapshot();
  update(state => state.consumer2.value++);
  expect(component.toJSON()).toMatchSnapshot();
});

test("Should re-rendering only part of consumers if nested.", () => {
  const { Provider, update, select } = define<State>();
  const renderCount = { consumer1: 0, consumer2: 0 };
  const Consumer1 = select(state => state.consumer1);
  const Consumer2 = select(state => state.consumer2);
  const Test1 = () => (
    <Consumer1>
      {state => (
        <div>
          {String(state.value)}
          {String(renderCount.consumer1++)}
        </div>
      )}
    </Consumer1>
  );
  const Test2 = () => (
    <Consumer2>
      {state => (
        <div>
          {String(state.value)}
          {String(renderCount.consumer2++)}
        </div>
      )}
    </Consumer2>
  );
  const component = create(
    <Provider state={state}>
      <Test1 />
      <Test2 />
    </Provider>
  );
  update(state => state.consumer1.value++);
  expect(component.toJSON()).toMatchSnapshot();
  update(state => state.consumer2.value++);
  expect(component.toJSON()).toMatchSnapshot();
});
