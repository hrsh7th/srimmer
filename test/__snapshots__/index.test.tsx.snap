// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should re-rendering only part of consumers if nested. 1`] = `
Array [
  <div>
    1
    1
  </div>,
  <div>
    0
    0
  </div>,
]
`;

exports[`Should re-rendering only part of consumers if nested. 2`] = `
Array [
  <div>
    1
    1
  </div>,
  <div>
    1
    1
  </div>,
]
`;

exports[`Should re-rendering only part of consumers. 1`] = `
Array [
  <div>
    1
    1
  </div>,
  <div>
    0
    0
  </div>,
]
`;

exports[`Should re-rendering only part of consumers. 2`] = `
Array [
  <div>
    1
    1
  </div>,
  <div>
    1
    1
  </div>,
]
`;
